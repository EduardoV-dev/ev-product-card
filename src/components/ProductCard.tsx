import React, { createContext, CSSProperties } from 'react';
import {
  IInitialValues,
  IOnChangeArgs,
  IProductCardHandlers,
  IProductContextProps,
  Product,
} from '../interfaces/interfaces';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as IProductContextProps);
const { Provider } = ProductContext;

export interface IProps {
  // children?: ReactElement | ReactElement[];
  children: (args: IProductCardHandlers) => JSX.Element;
  className?: string;
  initialValues?: IInitialValues;
  onChange?: (args: IOnChangeArgs) => void;
  product: Product;
  style?: CSSProperties;
  value?: number;
}

const ProductCard = ({
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value,
}: IProps): JSX.Element => {
  const { counter, increaseBy, isMaxCountReached, reset } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialValues?.maxCount,
      }}
    >
      <div
        className={`${styles.productCard} ${className ?? ''}`}
        {...{ style }}
      >
        {children({
          counter,
          isMaxCountReached,

          maxCount: initialValues?.maxCount || Infinity,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
