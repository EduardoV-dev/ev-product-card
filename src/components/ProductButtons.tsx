import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProps {
  className?: string;
  style?: CSSProperties;
}

const ProductButtons = ({ className = '', style }: IProps) => {
  const { counter, increaseBy, maxCount = Infinity } = useContext(
    ProductContext
  );

  const isMaxReached: () => boolean = useCallback(() => counter >= maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} {...{ style }}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        {' '}
        -{' '}
      </button>
      <div className={styles.countLabel}> {counter} </div>

      <button
        className={`${styles.buttonAdd} ${
          isMaxReached() ? styles.disabled : ''
        }`}
        disabled={isMaxReached()}
        onClick={() => increaseBy(1)}
      >
        {' '}
        +{' '}
      </button>
    </div>
  );
};

export default ProductButtons;
