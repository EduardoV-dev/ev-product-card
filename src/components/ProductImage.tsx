import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProps {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

const ProductImage = ({ className = '', img, style }: IProps) => {
  const { product } = useContext(ProductContext);

  const imgToShow: string = img ?? (product.img = '');

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Coffee Mug"
      {...{ style }}
    />
  );
};

export default ProductImage;
