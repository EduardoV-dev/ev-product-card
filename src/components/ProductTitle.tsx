import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProps {
  className?: string;
  style?: CSSProperties;
  title?: string;
}

const ProductTitle = ({ title, className = '', style }: IProps) => {
  const { product } = useContext(ProductContext);

  return (
    <span
      className={`${styles.productDescription} ${className}`}
      {...{ style }}
    >
      {title ?? product.title}
    </span>
  );
};

export default ProductTitle;
