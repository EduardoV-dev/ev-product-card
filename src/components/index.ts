import { IProductCardHOCProps } from '../interfaces/interfaces';
import ProductButtons from './ProductButtons';
import ProductCardHOC from './ProductCard';
import ProductImage from './ProductImage';
import ProductTitle from './ProductTitle';

export { default as ProductTitle } from './ProductTitle';
export { default as ProductImage } from './ProductImage';
export { default as ProductButtons } from './ProductButtons';

export const ProductCard: IProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
