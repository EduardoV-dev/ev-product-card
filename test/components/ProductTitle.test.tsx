import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../fixtures';

describe('<ProductTitle />', () => {
  it('Should render correctly with the personalized title', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <ProductTitle title="Custom Product" />}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("Should render with the product's name", () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
