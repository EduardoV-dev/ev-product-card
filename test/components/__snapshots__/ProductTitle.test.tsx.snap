// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTitle /> Should render correctly with the personalized title 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Custom Product
  </span>
</div>
`;

exports[`<ProductTitle /> Should render with the product's name 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Coffee Mug - Sin Imagen
  </span>
</div>
`;
